# 2장 2강: 분산분석(ANOVA)과 사후 검정 이론 — 실습문제

## 실습 목표

- 세 집단 이상의 평균을 일원배치 분산분석으로 비교할 수 있다.
- F통계량과 p-value를 이용하여 전체 집단 차이를 판단할 수 있다.
- 집단 간·집단 내 변동으로 ANOVA 표를 구성하고 해석할 수 있다.
- ANOVA가 유의할 때 Tukey HSD 사후 검정을 수행할 수 있다.
- 유의한 집단 쌍과 평균 차이의 크기를 근거로 차이 구조를 설명할 수 있다.

## 실습 환경 / 데이터

- Python
- NumPy, pandas
- scipy.stats
- statsmodels
- `ames_housing(1).csv`

| 컬럼 | 의미 |
|---|---|
| `SalePrice` | 주택 판매가격 |
| `OverallQual` | 주택의 전반적인 품질 점수 |
| `KitchenQual` | 주방 품질 |

> 모든 검정의 유의수준은 `α = 0.05`입니다.  
> 표본은 지정된 `random_state`로 추출하여 결과를 재현합니다.

## 실습 준비

1. 필요한 라이브러리를 불러오세요.
2. Ames Housing 데이터를 `df`에 불러오세요.
3. 데이터 크기, 결측치 수, 컬럼명과 상위 5개 행을 확인하세요.

In [1]:
# 실습 준비 코드를 작성하세요.
import pandas as pd
df = pd.read_csv(r'C:\Machine_Learning_worksheet\Data\ames_housing.csv', encoding= 'utf-8-sig')

---

## 필수 1. One-way ANOVA로 전체 평균 차이 확인

### 문제 1-1. 전반적인 품질 점수 5·6·7 집단 비교

#### 문제 설명

주택의 전반적인 품질 점수가 5점, 6점, 7점인 세 집단의 평균 판매가격을 비교하려고 합니다. 각 집단에서 20개씩 표본을 추출하고 일원배치 분산분석을 수행하세요.

#### 요구사항

1. `OverallQual`이 5, 6, 7인 각 집단의 `SalePrice`에서 `n=20`, `random_state=5`로 표본을 추출하세요.
2. 각 집단의 표본 수, 평균, 표준편차를 출력하세요.
3. 세 집단이 서로 다른 주택으로 구성된 독립집단임을 설명하세요.
4. 각 집단에 Shapiro-Wilk 정규성 검정을 수행하세요.
5. 세 집단에 Levene 등분산 검정을 수행하세요.
6. 다음 가설을 작성하세요.
   - H₀: 세 집단의 모집단 평균 판매가격은 모두 같다.
   - H₁: 적어도 한 집단의 모집단 평균 판매가격은 다르다.
7. `stats.f_oneway()`로 일원배치 ANOVA를 수행하세요.
8. F통계량과 p-value를 출력하고 전체 차이 유무를 판단하세요.
9. ANOVA 결과만으로 어느 집단끼리 다른지 알 수 있는지 설명하세요.

#### 해석 질문

**Q1.** 세 집단을 각각 t검정으로 반복 비교하면 어떤 문제가 발생하나요?  
**Q2.** F통계량은 어떤 두 변동의 비율인가요?  
**Q3.** ANOVA 결과 세 집단의 평균 판매가격에는 전체적으로 유의한 차이가 있나요?  
**Q4.** ANOVA 결과만으로 5점·6점·7점 중 어느 집단 쌍이 다른지 알 수 있나요?

#### 제출 결과

- 집단별 기술통계량과 가정 점검 결과
- 가설 설정
- F통계량과 p-value
- 전체 차이 판단
- 사후 검정 필요성 설명
- Q1~Q4 답변

In [ ]:
# 필수 1 코드를 작성하세요.

### 필수 1 답변 작성란

- **Q1.**
검정을 반복할 수록 전체 분석에서 한 번 이상 제 1종 오류가 발생 확률이 0.05보다 커지는 다중 비교 문제가 발생한다
- **Q2.**
집단 간 평균 차이를 나타내는 집단 간 변동을 같은 집단 내부의 개인차인 집단 ㅐ 변동으로 나눈 비율

- **Q3.**
있습니다. 따라서 적어도 한 집단의 모집단 평균 판매가격은 다르다.

- **Q4.**
알 수 없음, 구체적인 집단 쌍은 Tukey HSD와 같은 사후 검정으로 확인해야한다.


---

## 필수 2. ANOVA 표 구성과 Tukey HSD 사후 검정

### 문제 2-1. 품질 점수별 차이 구조 확인

#### 문제 설명

필수 1의 세 집단을 이용하여 ANOVA 표를 직접 구성하고, 전체 차이가 유의한 경우 Tukey HSD 사후 검정을 수행해 어느 품질 점수 집단끼리 차이가 있는지 확인하세요.

#### 요구사항

1. 필수 1의 세 집단을 하나의 `anova_df` 데이터프레임으로 결합하세요.
2. 전체 평균을 계산하세요.
3. 다음 값을 계산하여 ANOVA 표를 만드세요.
   - 집단 간 제곱합 `SS_between`
   - 집단 내 제곱합 `SS_within`
   - 집단 간·집단 내 자유도
   - 평균제곱 `MS_between`, `MS_within`
   - F통계량
4. 직접 계산한 F통계량이 `stats.f_oneway()` 결과와 일치하는지 확인하세요.
5. ANOVA p-value가 0.05보다 작을 때만 `pairwise_tukeyhsd()`를 실행하세요.
6. Tukey 결과에서 `reject=True`인 집단 쌍을 확인하세요.
7. 각 집단 평균을 이용하여 집단 쌍별 평균 차이를 계산하세요.
8. 어느 집단 쌍이 유의하며 차이가 가장 큰 집단 쌍은 무엇인지 해석하세요.

#### 해석 질문

**Q1.** ANOVA 표에서 `SS_between`과 `SS_within`은 각각 무엇을 의미하나요?  
**Q2.** Tukey HSD의 `reject=True`는 무엇을 의미하나요?  
**Q3.** 어느 품질 점수 집단 쌍에서 유의한 차이가 확인되나요?  
**Q4.** 평균 판매가격 차이가 가장 큰 집단 쌍은 무엇이며 차이는 얼마인가요?

#### 제출 결과

- ANOVA 표
- F통계량 대조 결과
- Tukey HSD 결과
- 유의한 집단 쌍
- 집단 쌍별 평균 차이
- Q1~Q4 답변

In [ ]:
# 필수 2 코드를 작성하세요.

### 필수 2 답변 작성란

- **Q1.**
ss_between 은 집단 평균들이 전체 평균에서 벗어난 집단 간 변동 값
ss_within은 각 관측값이 소속 집단 평균에서 벗어난 집단 내 변동
- **Q2.**
-> 다중 비교 오류를 조정한 뒤에도 해당 두 집단의 평균이 같다는 귀무가설을 기각하겟다
- **Q3.**

- **Q4.**

---

## 과제. 주방 품질에 따른 판매가격 차이 분석

### 문제 3-1. 주방 품질 `Ex`·`Gd`·`TA` 집단 비교

#### 문제 설명

주방 품질이 `Ex`, `Gd`, `TA`인 세 집단의 평균 판매가격을 비교합니다. 각 집단에서 20개씩 표본을 추출한 뒤 ANOVA와 Tukey HSD를 순서대로 적용하세요.

> 필수 문제에서 학습한 전체 검정→사후 검정 절차를 새로운 집단 변수에 적용하는 과제입니다.

#### 요구사항

1. `KitchenQual`이 `Ex`, `Gd`, `TA`인 각 집단의 `SalePrice`에서 `n=20`, `random_state=18`로 표본을 추출하세요.
2. 세 집단의 표본 수와 평균을 출력하세요.
3. 정규성과 등분산성을 확인하세요.
4. 일원배치 ANOVA를 수행하고 F통계량과 p-value를 출력하세요.
5. ANOVA가 유의한 경우에만 Tukey HSD 사후 검정을 수행하세요.
6. Tukey 결과에서 유의한 집단 쌍을 확인하세요.
7. 각 집단 쌍의 평균 판매가격 차이를 계산하세요.
8. 어느 집단 쌍의 차이가 가장 큰지 포함하여 주방 품질별 차이 구조를 해석하세요.

#### 해석 질문

**Q1.** ANOVA 결과 세 집단의 평균에는 전체적으로 유의한 차이가 있나요?  
**Q2.** 사후 검정은 어떤 조건에서 수행하나요?  
**Q3.** Tukey HSD에서 유의한 차이가 확인된 집단 쌍은 무엇인가요?  
**Q4.** 평균 판매가격 차이가 가장 큰 집단 쌍은 무엇이며 차이는 얼마인가요?

#### 제출 결과

- 집단별 기술통계량과 가정 점검
- ANOVA 결과
- Tukey HSD 결과
- 유의한 집단 쌍과 평균 차이
- 최종 해석
- Q1~Q4 답변

In [2]:
# 과제 코드를 작성하세요.

import numpy as np
import pandas as pd
from scipy import stats
from statsmodels.stats.multicomp import pairwise_tukeyhsd

# 1) 세 집단 표본 추출
groups = {}
for q in ['Ex', 'Gd', 'TA']:
    groups[q] = df.loc[df['KitchenQual'] == q, 'SalePrice'].dropna().sample(20, random_state=18)

ex, gd, ta = groups['Ex'], groups['Gd'], groups['TA']

# 2) 표본 수와 평균
for q, s in groups.items():
    print(f"{q}: n = {len(s)}, 평균 = {s.mean():.4f}")

# 3) 정규성 (Shapiro-Wilk)
print()
for q, s in groups.items():
    sw = stats.shapiro(s)
    print(f"[정규성] {q}: W = {sw.statistic:.4f}, p = {sw.pvalue:.4f} "
          f"→ {'충족' if sw.pvalue > 0.05 else '위배'}")

# 3) 등분산성 (Levene)
lev = stats.levene(ex, gd, ta)
print(f"\n[등분산성] Levene: W = {lev.statistic:.4f}, p = {lev.pvalue:.4f} "
      f"→ {'충족' if lev.pvalue > 0.05 else '위배'}")

# 4) 일원배치 ANOVA
f_stat, p_anova = stats.f_oneway(ex, gd, ta)
print(f"\nF-통계량 = {f_stat:.4f}")
print(f"p-value  = {p_anova:.4f}  (지수: {p_anova:.4e})")

# 5) 유의한 경우에만 Tukey HSD
if p_anova <= 0.05:
    print("\nANOVA 유의 → Tukey HSD 수행\n")
    data = pd.concat([ex, gd, ta])
    labels = ['Ex']*20 + ['Gd']*20 + ['TA']*20
    tukey = pairwise_tukeyhsd(endog=data, groups=labels, alpha=0.05)
    print(tukey)
else:
    print("\nANOVA 유의하지 않음 → 사후검정 생략")

# 7) 집단 쌍별 평균 차이
print()
pairs = [('Ex','Gd'), ('Ex','TA'), ('Gd','TA')]
for a, b in pairs:
    d = groups[a].mean() - groups[b].mean()
    print(f"{a} - {b} = {d:,.4f}")

Ex: n = 20, 평균 = 313983.0500
Gd: n = 20, 평균 = 188835.0000
TA: n = 20, 평균 = 137486.6000

[정규성] Ex: W = 0.9669, p = 0.6875 → 충족
[정규성] Gd: W = 0.9723, p = 0.8024 → 충족
[정규성] TA: W = 0.9519, p = 0.3973 → 충족

[등분산성] Levene: W = 2.5550, p = 0.0866 → 충족

F-통계량 = 54.8000
p-value  = 0.0000  (지수: 5.3052e-14)

ANOVA 유의 → Tukey HSD 수행

      Multiple Comparison of Means - Tukey HSD, FWER=0.05       
group1 group2  meandiff  p-adj     lower        upper     reject
----------------------------------------------------------------
    Ex     Gd -125148.05    0.0 -166883.2108  -83412.8892   True
    Ex     TA -176496.45    0.0 -218231.6108 -134761.2892   True
    Gd     TA   -51348.4 0.0122  -93083.5608   -9613.2392   True
----------------------------------------------------------------

Ex - Gd = 125,148.0500
Ex - TA = 176,496.4500
Gd - TA = 51,348.4000


### 과제 답변 작성란

**Q1.** ANOVA 결과 세 집단의 평균에는 전체적으로 유의한 차이가 있나요?  
ANOVA 결과 F= 54.8
p = 5.3053e^-14 보다 작아 귀무 가설을 기각한다. 세 집단중에서 한쌍에서 평균 차이가 존재한다.


**Q2.** 사후 검정은 어떤 조건에서 수행하나요?  
ANOVA가 유의 할 때만 수행한다. ANOVA가 어딘가 차이가 있다는거는 알뿐 어느 곳인지는 말하진 않고 t 검정을 3 번 반복하면 1종오류가 약 14% 까지 누적되기 때문에 Tukey HSD로 보정해 비교한다. 


**Q3.** Tukey HSD에서 유의한 차이가 확인된 집단 쌍은 무엇인가요?  
세 결과 모두 유의한다. TUKey HSD 결과 EX-GD, EX-TA , GD-TA 전부 reject이  True. 그렇기에 서로 구별되는 가격대를 형성한다. 


**Q4.** 평균 판매가격 차이가 가장 큰 집단 쌍은 무엇이며 차이는 얼마인가요?

평균이 가장 큰집단 쌍은 EX-TA이고 176,496.45 달러로 제일 크다. 

---

## 실습 마무리

1. 세 집단 이상을 t검정으로 반복 비교하면 왜 제1종 오류가 커지나요?
각 검정마다 위양성 가능성이 있기 때문에 비교 횟수가 늘어날수록 전체 분석에서 한번이상 잘못 기각할 확률이 누적이다.

2. ANOVA의 귀무가설과 대립가설은 무엇인가요?
귀무가설은 모든 집단의 모집단과 평균이 같다, 대립가설은 적어도 한 집단의 평균이 다르다라는것
3. F통계량이 크다는 것은 무엇을 의미하나요?
집단 간 차이 집단 내 오차보다 크다는것을 의미한다. 즉 집단간 평균 차이가 단순한 오차나 우연에 의한 것이 아니라 실제 집단간에 유의미한 차이가 존재할 가능성이 높음을 뜻한다
4. ANOVA가 유의하더라도 사후 검정이 필요한 이유는 무엇인가요?
ANOVA 결과는 적어도 한 집단은 다르다는 사실만 알려줄 뿐, 정확히 어떤 집단 간에 차이가 있는지는 알려주지 않기 때문이다

5. Tukey HSD 결과에서 어떤 항목을 확인해야 하나요?
P-value :0.05보다 작은지 확인하여 두 집단 간 차이가 유의미 한지 판별합니다.
diff : 두집단 간 차이 크기와 방향을 확인한다
 신뢰구간: 구간 안에 0이 포함되어 있지 않은지 확인합니다.
 